[Debugger] Provide SolutionItem to DebuggerSession mapping

mrward opened the pull request:
The DebuggerService now has two methods that can map from a SolutionItem
to a DebuggerSession or the other way around. The ExecutionCommand now
has a SolutionItem property which is used by the DebuggerService to
create this mapping from DebuggerSession to SolutionItem.

Fixes VSTS #1012614 - Improvements to DebuggerSession for XAML Hot Reload

mrward committed:
[Debugger] Provide project to DebuggerSession mapping

The DebuggerService now has two methods that can map from an IRunTarget
to a DebuggerSession or the other way around. The ExecutionCommand now
has an IRunTarget property which is used by the DebuggerService to
create this mapping from DebuggerSession to IRunTarget. IRunTarget
is implemented by SolutionItem.

Fixes VSTS #1012614 - Improvements to DebuggerSession for XAML Hot Reload

mrward commented:
I have some concerns about thread safety in DebuggerService. No locks are used when accessing the sessions dictionary. Some methods check the GUI thread is used but others do not and it looks like they would not always be called on the UI thread. Whilst ToArray is being used in some places I am not sure that is enough to solve the problem.

Using a ConcurrentDictionary might be a simple solution here, however the ordering of the items in a ConcurrentDictionary is not guaranteed and behaves different to the Dictionary type.

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core.Execution/ExecutionCommand.cs:
```diff
@@ -46,5 +46,11 @@ public abstract class ExecutionCommand
+		/// Solution item associated with this execution command. This allows the DebuggerSession to be
+		/// associated with a Solution item.
+		/// </summary>
+		public SolutionItem SolutionItem { get; set; }
```
Should this be [IRunTarget](http://source.monodevelop.com/#MonoDevelop.Core/MonoDevelop.Projects/IRunTarget.cs,bf4f319052985c6c)?

drasticactions reviewed (approved):
lgtm! 👍

jstedfast commented on main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/DebuggingService.cs:
```diff
@@ -149,6 +149,26 @@ public static IExecutionHandler GetExecutionHandler ()
+
+		public static IRunTarget GetRunTarget (DebuggerSession session)
+		{
+			foreach (KeyValuePair<DebuggerSession, SessionManager> item in sessions.ToArray ()) {
```
Couldn't this implementation just use `sessions.TryGetValue (session, out var runTarget)`?

mrward replied:
Indeed it could. Must have just based this code on the code above. Still a bit worried about thread safety here. I am assuming TryGetValue might not be happy if another thread is adding/removing sessions at the same tim.

Therzok commented on main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/DebuggingService.cs:
```diff
@@ -149,6 +149,26 @@ public static IExecutionHandler GetExecutionHandler ()
 
+		public static DebuggerSession GetSession (IRunTarget runTarget)
+		{
+			foreach (KeyValuePair<DebuggerSession, SessionManager> item in sessions.ToArray ()) {
```
This ToArray() should not be needed. Is there any thread-safe way to modify sessions? Otherwise, nothing is executing code that could potentially modify the collection on the current running thread here.

mrward replied:
That was my worry. There is no locking done in the DebuggerService. Some code seems to assuming that the UI thread is being used, but then you see Runtime.RunInMainThread being called directly. There is also no checks made to ensure the main UI thread is being used. So not sure what the locking strategy should be in the DebuggerService.

ConcurrentDictionary is an option but then the order of the keys will be different which would affect the CurrentSession property which uses `sessions.Keys.FirstOrDefault()`.

mrward committed:
[Debugger] Make DebuggingService dictionaries thread safe.

The sessions dictionary and the nextStatementLocations dictionary
were accessed from multiple threads. Changed these dictionaries to
be ConcurrentDictionary so their use is thread safe.
